Web Extensions: Parse getRecent() bookmark function

akutira-umich opened the pull request:
#### 66634381c1e4ac756c042f55015e98a85dd61b04
<pre>
Web Extensions: Parse getRecent() bookmark function
<a href="https://rdar.apple.com/153924872">rdar://153924872</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=294766">https://bugs.webkit.org/show_bug.cgi?id=294766</a>

Reviewed by Timothy Hatcher.

Added Parsing for getRecent that uses mock BookmarkTreeNodes created by Create()
to test its functionality. Added tests for this as well.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::WebExtensionAPIBookmarks::createDictionaryFromNode):
(WebKit::WebExtensionAPIBookmarks::createBookmark):
(WebKit::WebExtensionAPIBookmarks::getRecent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIBookmarks.idl:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPICheckgetRecent)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIMockNodeWithgetRecent)):

Canonical link: <a href="https://commits.webkit.org/296640@main">https://commits.webkit.org/296640@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66e36a3b60c0e5cdccce0b2c7071dba4a5fb039a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109173 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28831 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19258 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114380 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/59451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111136 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29512 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37398 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/82964 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/59451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112121 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/23482 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98331 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63413 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/22872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16474 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/59029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92849 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16516 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117495 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36217 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/26774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91980 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36589 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94594 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91786 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36707 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14457 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/32066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17614 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36113 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/41608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35799 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39133 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37490 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
```diff
@@ -29,11 +29,24 @@
+private:
+
+    struct MockBookmarkNode {
+        std::string id;
```
These should all either be `NSString` or `WTF::String`, but I think we tend to shy away from `std::string`

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -47,7 +88,27 @@
+    for (size_t i = 0; i < countToReturn; ++i) {
+        NSDictionary *nodeDict = createDictionaryFromNode(sortedBookmarks[i]);
+        [resultArray addObject:nodeDict];
+    }
```
Could probably use `std::reduce` here instead of the loop, but that's not super important

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
 
+NSDictionary* WebExtensionAPIBookmarks::createDictionaryFromNode(const MockBookmarkNode& node)
+{
+    double msSinceEpoch = (node.dateAdded + kCFAbsoluteTimeIntervalSince1970) * 1000.0;
```
You should use `toImpl()` from `CocoaHelpers.h`.

```c++
WallTime toImpl(NSDate *);
```

Then `toImpl(node.dateAdded).secondsSinceEpoch().milliseconds()`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
 
 namespace WebKit {
 
+NSDictionary* WebExtensionAPIBookmarks::createDictionaryFromNode(const MockBookmarkNode& node)
```
```suggestion
NSDictionary *WebExtensionAPIBookmarks::createDictionaryFromNode(const MockBookmarkNode& node)
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
+        @"id": @(node.id.c_str()),
+        @"title": @(node.title.c_str()),
+        @"url": @(node.url.c_str()),
+        @"dateAdded": @(static_cast<long long>(msSinceEpoch)),
```
```suggestion
        @"dateAdded": @(toImpl(node.dateAdded).secondsSinceEpoch().milliseconds()),
```

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
```diff
@@ -29,11 +29,24 @@
+        CFAbsoluteTime dateAdded;
+    };
+
+    std::vector<MockBookmarkNode> m_mockBookmarks;
```
This should probably be `WTF::Vector<MockBookmarkNode>`

xeenon replied:
```suggestion
    Vector<MockBookmarkNode> m_mockBookmarks;
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
+        *outExceptionString = @"Invalid argument: 'bookmark' details must be an object.";
+        callback->reportError(*outExceptionString);
+        return;
+    }
```
This can be removed, it is checked at the bindings level.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
+        *outExceptionString = @"For testing, 'id' (string), 'url' (string), and 'dateAdded' (number) must be provided.";
+        callback->reportError(*outExceptionString);
+        return;
+    }
```
This should be using `validateDictionary()`. See other APIs for how it is used.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
+    newNode.title = [titleValue isKindOfClass:[NSString class]] ? [static_cast<NSString*>(titleValue) UTF8String] : "";
+
+    double msSinceEpoch = [static_cast<NSNumber*>(dateAddedValue) doubleValue];
+    newNode.dateAdded = (msSinceEpoch / 1000.0) - kCFAbsoluteTimeIntervalSince1970;
```
See other comment about `toImpl()`.

AriYoung00 replied:
After switching `dateAdded` to `NSDate *`:
```suggestion
    newNode.dateAdded = [NSDate dateWithTimeIntervalSince1970:msSinceEpoch];
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
+
+    MockBookmarkNode newNode;
+    newNode.id = [static_cast<NSString*>(idValue) UTF8String];
+    newNode.url = [static_cast<NSString*>(urlValue) UTF8String];
```
`UTF8String` is not safe to use, the `char *` will be a dangling pointer. `MockBookmarkNode` should use `WTF::String`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
+    newNode.url = [static_cast<NSString*>(urlValue) UTF8String];
+
+    id titleValue = bookmark[@"title"];
+    newNode.title = [titleValue isKindOfClass:[NSString class]] ? [static_cast<NSString*>(titleValue) UTF8String] : "";
```
Can be cleaned up with `validateDictionary()` above. We should also avoid `isKindOfClass:`.

AriYoung00 replied:
After switching `title` to `WTF::String`:
```suggestion
    newNode.title = [titleValue isKindOfClass:[NSString class]] ? dynamic_objc_cast<NSString>(titleValue) : ""_s;
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
+
+    m_mockBookmarks.push_back(newNode);
+    NSDictionary* resultNode = createDictionaryFromNode(newNode);
+    callback->call(resultNode);
```
```suggestion

    callback->call(createDictionaryFromNode(newNode));
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -47,7 +88,27 @@
+        NSString *errorMessage = @"Invalid argument: 'numberOfItems' must be at least 1.";
+        if (outExceptionString)
+            *outExceptionString = errorMessage;
+        callback->reportError(errorMessage);
```
You should match Chrome/Firefox here. You either want to throw an exception or `reportError()`, not both.

Also use `toErrorString()` for the error message. See other APIs for how it is used.

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
+    }
+
+    MockBookmarkNode newNode;
+    newNode.id = [static_cast<NSString*>(idValue) UTF8String];
```
After you switch to `WTF::String` for `id` and `url`:
```suggestion
    newNode.id = dynamic_objc_cast<NSString>(idValue);
```
and similarly below for `url`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
```diff
@@ -29,11 +29,24 @@
+        std::string id;
+        std::string title;
+        std::string url;
+        CFAbsoluteTime dateAdded;
```
Prefer `WTF::WallTime`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
```diff
@@ -29,11 +29,24 @@
+    struct MockBookmarkNode {
+        std::string id;
+        std::string title;
+        std::string url;
```
Prefer `WTF::String`.

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
+    newNode.dateAdded = (msSinceEpoch / 1000.0) - kCFAbsoluteTimeIntervalSince1970;
+
+    m_mockBookmarks.push_back(newNode);
+    NSDictionary* resultNode = createDictionaryFromNode(newNode);
```
```suggestion
    RetainPtr resultNode = createDictionaryFromNode(newNode);
```

xeenon replied:
`RetainPtr` is not needed since we use ARC.
```suggestion
    auto *resultNode = createDictionaryFromNode(newNode);
```

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
+    id titleValue = bookmark[@"title"];
+    newNode.title = [titleValue isKindOfClass:[NSString class]] ? [static_cast<NSString*>(titleValue) UTF8String] : "";
+
+    double msSinceEpoch = [static_cast<NSNumber*>(dateAddedValue) doubleValue];
```
```suggestion
    double msSinceEpoch = dynamic_objc_cast<NSNumber>(dateAddedValue).doubleValue;
```

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
 void WebExtensionAPIBookmarks::createBookmark(NSDictionary *bookmark, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
 {
-    callback->reportError(@"unimplemented");
+    if (![bookmark isKindOfClass:[NSDictionary class]]) {
```
```suggestion
    if (![bookmark isKindOfClass:NSDictionary.class]) {
```

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
+    id urlValue = bookmark[@"url"];
+    id dateAddedValue = bookmark[@"dateAdded"];
+
+    if (![idValue isKindOfClass:[NSString class]] || ![urlValue isKindOfClass:[NSString class]] || ![dateAddedValue isKindOfClass:[NSNumber class]]) {
```
```suggestion
    if (![idValue isKindOfClass:NSString.class] || ![urlValue isKindOfClass:NSString.class] || ![dateAddedValue isKindOfClass:NSNumber.class]) {
```

And similarly for any other calls to `class`. In general, we prefer to use dot syntax for property access.

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -35,9 +35,50 @@
+    double msSinceEpoch = (node.dateAdded + kCFAbsoluteTimeIntervalSince1970) * 1000.0;
+
+    return @{
+        @"id": @(node.id.c_str()),
```
After switching `id` to `WTF::String`:
```suggestion
        @"id": node.id.createNSString().get(),
```
and similarly for other string fields

AriYoung00 replied:
Actually, just `node.id.get()` might work here? Not totally sure, but worth a shot

AriYoung00 reviewed (changes requested):
Requesting changes based on previous comments

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -30,14 +30,53 @@
+    if (!validateDictionary(bookmark, @"bookmark", requiredKeys, types, &exceptionString)) {
+        callback->reportError(exceptionString);
+        return;
+    }
```
Input validation is always an exception.
```suggestion
    if (!validateDictionary(bookmark, @"bookmark", requiredKeys, types, outExceptionString))
        return;

```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -30,14 +30,53 @@
 {
-    callback->reportError(@"unimplemented");
+    id idValue = bookmark[@"id"];
+    id dateAddedValue = bookmark[@"dateAdded"];
```
This should be after validation, then you can assume the correct types.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -30,14 +30,53 @@
+    id idValue = bookmark[@"id"];
+    id dateAddedValue = bookmark[@"dateAdded"];
+
+    NSArray *requiredKeys = @[@"id", @"url", @"dateAdded"];
```
```suggestion
    NSArray *requiredKeys = @[ @"id", @"url", @"dateAdded" ];
```
These keys should be constants at the top of the time. See other APIs for examples.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -30,14 +30,53 @@
+    }
+    MockBookmarkNode newNode;
+    newNode.id = dynamic_objc_cast<NSString>(idValue);
+    newNode.url = dynamic_objc_cast<NSString>(idValue);
```
No need for the cast after validation. Also you use `idValue` for the URL.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -30,14 +30,53 @@
+        @"url": [NSString class],
+        @"dateAdded": [NSNumber class],
+        @"title": [NSString class]
+    };
```
```suggestion
    static NSDictionary<NSString *, id> *types = @{
        @"id": NSString.class,
        @"url": NSString.class,
        @"dateAdded": NSNumber.class,
        @"title": NSString.class
    };
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -30,14 +30,53 @@
+    newNode.url = dynamic_objc_cast<NSString>(idValue);
+
+    id titleValue = bookmark[@"title"];
+    newNode.title = [titleValue isKindOfClass:[NSString class]] ? WTF::String(dynamic_objc_cast<NSString>(titleValue)) : ""_s;
```
`isKindOfClass:` is not needed after validation.
```suggestion
    newNode.title = bookmark[titleKey];
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -30,14 +30,53 @@
+    newNode.title = [titleValue isKindOfClass:[NSString class]] ? WTF::String(dynamic_objc_cast<NSString>(titleValue)) : ""_s;
+
+    double msSinceEpoch = dynamic_objc_cast<NSNumber>(dateAddedValue).doubleValue;
+    newNode.dateAdded = WTF::WallTime::fromRawSeconds(msSinceEpoch / 1000.0);
```
```suggestion
    newNode.dateAdded = WallTime::fromRawSeconds(Seconds::fromMilliseconds(objectForKey<NSDate>(bookmark, dateAddedKey).doubleValue).value());
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -30,14 +30,53 @@
+    newNode.dateAdded = WTF::WallTime::fromRawSeconds(msSinceEpoch / 1000.0);
+
+    m_mockBookmarks.append(newNode);
+    callback->call(createDictionaryFromNode(newNode));
```
```suggestion
    m_mockBookmarks.append(WTFMove(newNode));

    callback->call(createDictionaryFromNode(newNode));
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -47,7 +86,25 @@
+    if (numberOfItems < 1) {
+        NSString *errorMessage = @"Invalid argument: 'numberOfItems' must be at least 1.";
+        if (outExceptionString)
+            *outExceptionString = errorMessage;
```
Need to use `toErrorString()` and always return. You can safely assume `outExceptionString` is not null here.
```suggestion
        *outExceptionString = toErrorString(nullString(), numberOfItemsKey, @"it must be at least 1").createNSString().autorelease();
        return;
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -47,7 +86,25 @@
+    for (size_t i = 0; i < countToReturn; ++i) {
+        NSDictionary *nodeDict = createDictionaryFromNode(sortedBookmarks[i]);
+        [resultArray addObject:nodeDict];
+    }
```
```suggestion
    size_t countToReturn = std::min(static_cast<size_t>(numberOfItems), sortedBookmarks.size());
    auto *resultArray = [NSMutableArray arrayWithCapacity:countToReturn];
    for (size_t i = 0; i < countToReturn; ++i)
        [resultArray addObject:createDictionaryFromNode(sortedBookmarks[i])];
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
```diff
@@ -34,6 +34,18 @@ namespace WebKit {
+        WTF::String id;
+        WTF::String title;
+        WTF::String url;
+        WTF::WallTime dateAdded;
```
```suggestion
        String id;
        String title;
        String url;
        WallTime dateAdded;
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -154,16 +154,54 @@
 {
     auto *script = @[
-        @"browser.test.assertThrows(() => browser.bookmarks.getChildren(123), /The 'id' value is invalid, because a string is expected/i)",
-        @"browser.test.assertThrows(() => browser.bookmarks.getChildren({}), /The 'id' value is invalid, because a string is expected/i)",
```
You can/should just rename the parameter name to `id` in the IDL file. `bookmarkIdentifier` is redundant.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -30,14 +30,51 @@
+    id dateAddedValue = bookmark[@"dateAdded"];
+    MockBookmarkNode newNode;
+    newNode.id = idValue;
+    newNode.url = urlValue;
```
```suggestion
    MockBookmarkNode newNode;
    newNode.id = bookmark[@"id"];
    newNode.url = bookmark[@"url"];
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -30,14 +30,51 @@
+        @"id": NSString.class,
+        @"url": NSString.class,
+        @"dateAdded": NSNumber.class,
+        @"title": NSString.class
```
These should still use constants defined at the top of the file.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -30,14 +30,54 @@
+static NSString * const WebExtensionBookmarkIDKey = @"id";
+static NSString * const WebExtensionBookmarkURLKey = @"url";
+static NSString * const WebExtensionBookmarkTitleKey = @"title";
+static NSString * const WebExtensionBookmarkDateAddedKey = @"dateAdded";
```
```suggestion
static NSString * const idKey = @"id";
static NSString * const urlKey = @"url";
static NSString * const titleKey = @"title";
static NSString * const dateAddedKey = @"dateAdded";
static NSString * const typeKey = @"type";
static NSString * const bookmarkKey = @"bookmark";
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -30,14 +30,54 @@
+        @"title": node.title.createNSString().get(),
+        @"url": node.url.createNSString().get(),
+        @"dateAdded": @(node.dateAdded.secondsSinceEpoch().milliseconds()),
+        @"type": @"bookmark"
```
```suggestion
        idKey: node.id.createNSString().get(),
        titleKey: node.title.createNSString().get(),
        urlKey: node.url.createNSString().get(),
        dateAddedKey: @(node.dateAdded.secondsSinceEpoch().milliseconds()),
        typeKey: bookmarkKey
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -30,14 +30,54 @@
+        WebExtensionBookmarkURLKey: [NSString class],
+        WebExtensionBookmarkDateAddedKey: [NSNumber class],
+        WebExtensionBookmarkTitleKey: [NSString class]
+    };
```
Use constants defined above here. `NSString.class`, `NSNumber.class`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -30,14 +30,54 @@
+    MockBookmarkNode newNode;
+    newNode.id = bookmark[@"id"];
+    newNode.url = bookmark[@"url"];
+    id dateAddedValue = bookmark[@"dateAdded"];
```
No need for a local. Should use `auto *` instead of `id`.

akutira-umich committed:
Web Extensions: Parse getRecent() bookmark function
rdar://153924872
https://bugs.webkit.org/show_bug.cgi?id=294766

Reviewed by Timothy Hatcher.

Added Parsing for getRecent that uses mock BookmarkTreeNodes created by Create()
to test its functionality. Added tests for this as well.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::WebExtensionAPIBookmarks::createDictionaryFromNode):
(WebKit::WebExtensionAPIBookmarks::createBookmark):
(WebKit::WebExtensionAPIBookmarks::getRecent):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIBookmarks.idl:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPICheckgetRecent)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIMockNodeWithgetRecent)):

Canonical link: https://commits.webkit.org/296640@main
